### ISSUE TITLE
[Refactor] Migrate AlexNet, MobileNetV1, GoogLeNet to use TrainingLoop

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ProjectOdyssey#2742]
## Problem

From audit #2602, three training scripts do NOT use the existing `TrainingLoop` infrastructure:
- `examples/alexnet-cifar10/train.mojo` - 552 lines
- `examples/mobilenetv1-cifar10/train.mojo` - 273 lines
- `examples/googlenet-cifar10/train.mojo` - 503 lines

## Current State

These scripts implement their own training epoch loops instead of using `shared/training/loops/training_loop.mojo`.

Scripts that DO use TrainingLoop:
- LeNet-5 ✓
- VGG-16 ✓
- ResNet-18 ✓

## Impact

- **Duplicated code:** ~80 lines of training loop logic across 3 scripts
- **Inconsistent behavior:** Different logging, batch handling
- **Maintenance burden:** Bug fixes need to be applied multiple times

## Suggested Fix

Update each script to use `TrainingLoop`:

```mojo
# Current (AlexNet example):
fn train_epoch(...) raises -> Float32:
    var num_samples = train_images.shape()[0]
    var num_batches = (num_samples + batch_size - 1) // batch_size
    for batch_idx in range(num_batches):
        # ... manual batch iteration

# Proposed:
from shared.training.loops import TrainingLoop
var training_loop = TrainingLoop(...)
training_loop.run_epoch(...)
```

## Related Issues

- Discovered in #2602 (Training Loop Audit)
- Related to #2597 (Training Loop Base Class)

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Now I have a comprehensive understanding of the codebase. Let me create the detailed implementation plan.

# Implementation Plan: Migrate AlexNet, MobileNetV1, GoogLeNet to TrainingLoop

## 1. Summary

This plan addresses Issue #2742 - migrating three training scripts to use the shared `TrainingLoop` infrastructure instead of implementing their own training epoch loops.

**Current State Analysis:**
- **AlexNet** (`examples/alexnet-cifar10/train.mojo`): 552 lines with custom `train_epoch()` function (lines 302-374)
- **MobileNetV1** (`examples/mobilenetv1-cifar10/train.mojo`): 273 lines with custom `train_epoch()` function (lines 62-133)
- **GoogLeNet** (`examples/googlenet-cifar10/train.mojo`): 503 lines with custom `train_epoch()` function (lines 87-340)

**Reference Implementation Pattern (from LeNet-5 and ResNet-18):**
The existing scripts that use `TrainingLoop` follow this pattern:
```mojo
from shared.training.loops import TrainingLoop

fn train_epoch(...) raises -> Float32:
    var loop = TrainingLoop(log_interval=100)
    
    fn compute_batch_loss(batch_images: ExTensor, batch_labels: ExTensor) raises -> Float32:
        # Model-specific forward pass and loss computation
        return Float32(loss_value)
    
    var avg_loss = loop.run_epoch_manual(
        train_images, train_labels,
        batch_size=batch_size,
        compute_batch_loss=compute_batch_loss,
        epoch=epoch,
        total_epochs=total_epochs,
    )
    return avg_loss
```

**Key Observations:**
1. `TrainingLoop.run_epoch_manual()` handles batch iteration, logging, and progress reporting
2. Model-specific logic is encapsulated in a `compute_batch_loss` closure
3. The closure captures model, learning_rate, momentum, and velocities from the outer scope
4. VGG-16's implementation has a bug - it tries to pass `compute_gradients(...)` directly instead of a closure

## 2. Step-by-Step Implementation Tasks

### Phase 1: AlexNet Migration

**Task 1.1: Update imports**
- Add: `from shared.training.loops import TrainingLoop`

**Task 1.2: Refactor `train_epoch()` function**
- Replace manual batch loop (lines 336-369) with `TrainingLoop.run_epoch_manual()`
- Wrap `compute_gradients()` call in a closure
- Maintain all existing functionality (forward pass, backward pass, SGD with momentum)

**Task 1.3: Update main training loop**
- Ensure `train_epoch()` signature remains compatible
- No changes to velocity initialization or model setup

**Estimated Line Changes:**
- Remove: ~35 lines (manual batch iteration logic)
- Add: ~20 lines (TrainingLoop integration)
- Net reduction: ~15 lines

### Phase 2: MobileNetV1 Migration

**Task 2.1: Update imports**
- Add: `from shared.training.loops import TrainingLoop`

**Task 2.2: Refactor `train_epoch()` function**
- Replace manual batch loop (lines 99-131) with `TrainingLoop.run_epoch_manual()`
- Wrap forward pass and loss computation in a closure
- Maintain batch normalization training mode

**Task 2.3: Update main training loop**
- Ensure compatibility with existing learning rate schedule

**Estimated Line Changes:**
- Remove: ~30 lines (manual batch iteration logic)
- Add: ~20 lines (TrainingLoop integration)
- Net reduction: ~10 lines

### Phase 3: GoogLeNet Migration

**Task 3.1: Update imports**
- Add: `from shared.training.loops import TrainingLoop`

**Task 3.2: Refactor `train_epoch()` function**
- Replace manual batch loop (lines 306-337) with `TrainingLoop.run_epoch_manual()`
- Wrap forward pass and loss computation in a closure
- Preserve Inception module documentation comments

**Task 3.3: Update main training loop**
- Ensure compatibility with existing learning rate schedule

**Estimated Line Changes:**
- Remove: ~30 lines (manual batch iteration logic)
- Add: ~20 lines (TrainingLoop integration)
- Net reduction: ~10 lines

### Phase 4: Verification and Cleanup

**Task 4.1: Ensure consistent logging**
- All three scripts should use `log_interval=100` (matching current behavior)
- Verify progress output format matches existing output

**Task 4.2: Code review for consistency**
- Verify closure captures work correctly for mutable model references
- Ensure velocity lists are properly captured in closures

## 3. Files to Modify

| File | Action | Changes |
|------|--------|---------|
| `examples/alexnet-cifar10/train.mojo` | Modify | Add TrainingLoop import, refactor `train_epoch()` |
| `examples/mobilenetv1-cifar10/train.mojo` | Modify | Add TrainingLoop import, refactor `train_epoch()` |
| `examples/googlenet-cifar10/train.mojo` | Modify | Add TrainingLoop import, refactor `train_epoch()` |

**No new files created** - this is a pure refactoring task.

## 4. Testing Approach

### 4.1 Pre-Migration Baseline
1. Record current output format from each script (first 2 epochs)
2. Document expected batch counts and loss values for reproducibility

### 4.2 Post-Migration Verification
1. **Syntax Verification**: `mojo build` each script to ensure compilation
2. **Output Consistency**: Compare logging format with baseline
3. **Functional Verification**: Run each script for 2 epochs and verify:
   - Batch iteration count matches
   - Loss values are computed correctly
   - Progress logging appears at correct intervals

### 4.3 Regression Testing
1. Run existing layerwise tests for each model:
   ```bash
   mojo test tests/models/test_alexnet_layers.mojo
   mojo test tests/models/test_mobilenetv1_layers.mojo  # if exists
   mojo test tests/models/test_googlenet_layers.mojo   # if exists
   ```

### 4.4 CI Validation
- Ensure pre-commit hooks pass (mojo format, markdownlint)
- Verify comprehensive-tests.yml workflow passes

## 5. Success Criteria

| Criterion | Verification Method |
|-----------|---------------------|
| All three scripts compile without errors | `mojo build` each script |
| No functional regression | Compare loss output for first 2 epochs |
| Consistent logging format | Visual inspection of output |
| Reduced code duplication | ~35 lines removed across all scripts |
| TrainingLoop import present | Code inspection |
| Pre-commit hooks pass | `pre-commit run --all-files` |
| Existing tests pass | Run layerwise tests |

## 6. Detailed Code Examples

### AlexNet Migration (Proposed)

**Before** (lines 302-374):
```mojo
fn train_epoch(
    mut model: AlexNet,
    train_images: ExTensor,
    train_labels: ExTensor,
    batch_size: Int,
    learning_rate: Float32,
    momentum: Float32,
    epoch: Int,
    total_epochs: Int,
    mut velocities: List[ExTensor],
) raises -> Float32:
    var num_samples = train_images.shape()[0]
    var num_batches = (num_samples + batch_size - 1) // batch_size
    var total_loss = Float32(0.0)

    print("Epoch [", epoch, "/", total_epochs, "]")

    for batch_idx in range(num_batches):
        # ... manual batch processing
        var batch_loss = compute_gradients(
            model, train_images, train_labels,
            learning_rate, momentum, velocities,
        )
        total_loss += batch_loss
        # ... manual progress logging
        break  # Demo mode

    var avg_loss = total_loss / Float32(num_batches)
    return avg_loss
```

**After**:
```mojo
fn train_epoch(
    mut model: AlexNet,
    train_images: ExTensor,
    train_labels: ExTensor,
    batch_size: Int,
    learning_rate: Float32,
    momentum: Float32,
    epoch: Int,
    total_epochs: Int,
    mut velocities: List[ExTensor],
) raises -> Float32:
    """Train for one epoch using TrainingLoop.

    Args:
        model: AlexNet model
        train_images: Training images (50000, 3, 32, 32)
        train_labels: Training labels (50000,)
        batch_size: Mini-batch size
        learning_rate: Learning rate for SGD
        momentum: Momentum factor
        epoch: Current epoch number (1-indexed)
        total_epochs: Total number of epochs
        velocities: Momentum velocities for each parameter

    Returns:
        Average loss for the epoch.
    """
    # Create training loop with progress logging every 100 batches
    var loop = TrainingLoop(log_interval=100)

    # Define compute_batch_loss closure that processes batches
    fn compute_batch_loss(
        batch_images: ExTensor, batch_labels: ExTensor
    ) raises -> Float32:
        # Compute gradients and update parameters
        return compute_gradients(
            model,
            batch_images,
            batch_labels,
            learning_rate,
            momentum,
            velocities,
        )

    # Run one epoch using the consolidated training loop
    var avg_loss = loop.run_epoch_manual(
        train_images,
        train_labels,
        batch_size=batch_size,
        compute_batch_loss=compute_batch_loss,
        epoch=epoch,
        total_epochs=total_epochs,
    )

    return avg_loss
```

### MobileNetV1 Migration (Proposed)

**After**:
```mojo
fn train_epoch(
    mut model: MobileNetV1,
    train_images: ExTensor,
    train_labels: ExTensor,
    batch_size: Int,
    learning_rate: Float32,
    momentum: Float32,
    epoch: Int,
) raises -> Float32:
    """Train for one epoch using TrainingLoop."""
    # Create training loop with progress logging every 100 batches
    var loop = TrainingLoop(log_interval=100)

    # Define compute_batch_loss closure
    fn compute_batch_loss(
        batch_images: ExTensor, batch_labels: ExTensor
    ) raises -> Float32:
        # Forward pass
        var logits = model.forward(batch_images, training=True)
        # Compute loss
        var loss = cross_entropy(logits, batch_labels)
        # Backward pass (see structure documentation in original)
        # Parameter updates (SGD with momentum)
        return loss

    # Total epochs not available in signature, pass 1 as placeholder
    var avg_loss = loop.run_epoch_manual(
        train_images,
        train_labels,
        batch_size=batch_size,
        compute_batch_loss=compute_batch_loss,
        epoch=epoch + 1,  # Convert 0-indexed to 1-indexed
        total_epochs=200,  # Default from main()
    )

    return avg_loss
```

### GoogLeNet Migration (Proposed)

**After**:
```mojo
fn train_epoch(
    mut model: GoogLeNet,
    train_images: ExTensor,
    train_labels: ExTensor,
    batch_size: Int,
    learning_rate: Float32,
    momentum: Float32,
    epoch: Int,
) raises -> Float32:
    """Train for one epoch using TrainingLoop."""
    # Create training loop with progress logging every 100 batches
    var loop = TrainingLoop(log_interval=100)

    # Define compute_batch_loss closure
    fn compute_batch_loss(
        batch_images: ExTensor, batch_labels: ExTensor
    ) raises -> Float32:
        # Forward pass (training mode)
        var logits = model.forward(batch_images, training=True)
        # Compute loss
        var loss = cross_entropy(logits, batch_labels)
        # Backward pass (see structure documentation above)
        # Parameter updates (see structure above)
        return loss

    var avg_loss = loop.run_epoch_manual(
        train_images,
        train_labels,
        batch_size=batch_size,
        compute_batch_loss=compute_batch_loss,
        epoch=epoch + 1,  # Convert 0-indexed to 1-indexed
        total_epochs=200,  # Default from main()
    )

    return avg_loss
```

## 7. Risk Assessment

| Risk | Likelihood | Impact | Mitigation |
|------|------------|--------|------------|
| Closure capture issues with mutable model | Medium | High | Test thoroughly; reference LeNet/ResNet patterns |
| Mojo closure semantics differ from expected | Low | Medium | Follow exact patterns from working implementations |
| Log output format changes | Low | Low | Cosmetic only; document if occurs |
| Performance regression | Low | Low | TrainingLoop uses same iteration pattern |

## 8. Dependencies

- Issue #2602 (Training Loop Audit) - Source of this refactoring request
- Issue #2597 (Training Loop Base Class) - Related infrastructure

## Resource Usage

- **Files Read**: 10
- **Tool Calls**: 10
- **Steps**: 10 (of 50 budget)

**Comment by mvillmow** [comment]
## Dependencies (AI-Analyzed)

**Priority**: P1

### Depends On
- #2597

### Analysis
Refactoring to use TrainingLoop requires the TrainingLoop base class to exist (#2597). This eliminates code duplication across 3 training scripts.

---
*Analyzed by Claude Code dependency analyzer*
